### ISSUE TITLE
chain jobs and use job.result for next job

### ISSUE THREAD

**Issue by benayat** [rq/rq#1503]
I chain several commands, with the "depends_on" parameter.
it works fine, until I try to use job1.result from the previous job, as a parameter for next one: 
   <span>    </span><p> job1 = current_app.task_queue.enqueue(background_task2, args = (2,))
   <span>    </span> job2 = current_app.task_queue.enqueue(background_task2, args = (job1.result,), depends_on=job1)</p><br/>
job1.result is always "None", unless I explicitly wait a few seconds (time.wait()), and run job1.result again.

is there a way to simply use the first job's result as a parameter for the next job?

**Comment by rpkak** [comment]
The problem is that if `job1` is enqueued, it is not executed yet and you don't have a result at this time. So at the time of the execution of the second line `job1.result` is `None`. To solve this you can fetch `job1` inside `job2` and I can open a PR to make this more elegant.

**Comment by benayat** [comment]
thanks for the answer! 
I'd do the PM myself if I was fluent enough...I'm just thinking loud - is it possible for rq to calculate the parameters only in the run-time of the request, instead of the enqueuing?

**Comment by rpkak** [comment]
> Is it possible for rq to calculate the parameters only in the runtime of the request, instead of the enqueuing?

No rq stores the parameters using a serialzer (default pickle) to redis. Fetching job results in the worker while preparing the next job would be a new feature

**Comment by rpkak** [comment]
Why did you close this issue?

**Comment by benayat** [comment]
I thought I was supposed to... I thought that since there is no option right now to get the results when running the new job, nothing can't be done for now...I'll just keep it open for now, until you can make a PR or until a solution is found. thanks.

**Comment by selwin** [comment]
So there are multiple ways we can go about this:
1. Use the new `on_success_callback` which automatically passes you the job instance, from which you can access `job.result`
2. Another way would be to have RQ's dependency feature optionally passing the some kind of extra data which lets you know which job triggers the dependent to run.

**Comment by mattchan-tencent** [comment]
@rpkak Yes, if you could please open a PR to make fetching `job1` within `job2` more elegant, it would be appreciated.

---

> So there are multiple ways we can go about this:
> 
> 1. Use the new `on_success_callback` which automatically passes you the job instance, from which you can access `job.result`
> 2. Another way would be to have RQ's dependency feature optionally passing the some kind of extra data which lets you know which job triggers the dependent to run.

@selwin there are a few downsides to those approaches IMO:

1. I think this means your job submission script needs to run for the entire duration of all the jobs correct? Something like this: https://stackoverflow.com/a/68346696
2. I tried something like this: 
```
job1 = q.enqueue(foo)
job2 = q.enqueue(bar, job1, depends_on=job1)
```
This fails because Python can't pickle a thread and its locks. 

---

The solution I've got so far is to fetch the job as @rpkak mentioned. Just in case others are looking:

```
# job
import redis
from rq.job import Job

def foo():
    return 5

def bar(jid):
    job = Job.fetch(jid, connection=Redis())
    return 2 * job.result

----
# job submission script snippet:

job1 = q.enqueue(foo)
job2 = q.enqueue(bar, job1.id, depends_on=job1)
```

This is a little bit ugly, since you have to reach into RQ internals in `bar`. You are also unable to determine the argument type in `bar` without knowing that `foo` was being called as a dependency. @rpkak It would be nicer if there was a decorator or a helper function that would automatically fetch `id` when enqueing job2 and fetch the result from Redis. That way the `bar` function could be written in the normal fashion and not require jumping around the codebase to understand.

**Comment by mattchan-tencent** [comment]
Also, if you implement a decorator, it might be nice to increase the 500s result timeout automatically to whatever the job timeout duration is. Fetching within 500s is reasonable, but if there's a lot of pressure on the queue, it's not uncommon for a job to require more time than that to start.

**Comment by selwin** [comment]
> This is a little bit ugly, since you have to reach into RQ internals in bar. You are also unable to determine the argument type in bar without knowing that foo was being called as a dependency. @rpkak It would be nicer if there was a decorator or a helper function that would automatically fetch id when enqueing job2 and fetch the result from Redis. That way the bar function could be written in the normal fashion and not require jumping around the codebase to understand.

I agree, I think we should definitely make reaching into parent jobs easier, something like this:
```python
for job in job2.dependencies:
    print(job.result)
```

**Comment by mattchan-tencent** [comment]
> I agree, I think we should definitely make reaching into parent jobs easier, something like this:
> 
> ```python
> for job in job2.dependencies:
>     print(job.result)
> ```

This is another alternative way to get access to the arguments in job (when paired with `get_current_job()`, but I don't think it's particular helpful since you still have to put that inside your `bar` code. 

I mean something more like this: 
``` python
# job

def fetch_results():
    def decorate(forig):
        def fnew(*args, **kwargs):
            args = [Job.fetch(a.id, connection=Redis()) for a in args]
            kwargs = {k:Job.fetch(v.id, connection=Redis()) for k,v in kwargs.items()}

            return forig(*args, **kwargs)
        fnew.func_name = forig.func_name
        return fnew
    return decorate

@fetch_results
def bar(x):
    return 2 * x

# ---
# submission

job1 = q.enqueue(foo)
q.enqueue(bar, job1, depends_on=job1)

```

Note that this won't work as-is. RQ will attempt to pickle the job when it enqueues and it will fail due to threading/mutex as mentioned earlier. It will require a few internal changes to RQ to work. Alternatively, users can write a wrapper for Queue.enqueue which strips the id out of the job and substitutes the arguments.

**Comment by selwin** [comment]
I have mixed feelings about the `fetch_results` decorator because:
1. You actually still need to update your code to use previous job results.
2. The decorator changes your function's signature in an implicit way that may break certain codes that access `args` and `kwargs`.

Maybe one thing we can do is when dependents are being [enqueued](https://github.com/rq/rq/blob/76ba690aafaa0fffbab1ee8f5db7c27c5327d792/rq/queue.py#L622), we check whether the callable accepts `rq_dependency_job_result` and pass an additional kwarg when executing the function.

**Comment by mattchan-tencent** [comment]
> 2. The decorator changes your function's signature in an implicit way that may break certain codes that access `args` and `kwargs`.

Yes, actually I had problems with this. I ended up using `functools.wraps` to fix it. RQ _really_ didn't like having an implicitly different function signature from the one being enqueued. 

> Maybe one thing we can do is when dependents are being [enqueued](https://github.com/rq/rq/blob/76ba690aafaa0fffbab1ee8f5db7c27c5327d792/rq/queue.py#L622), we check whether the callable accepts `rq_dependency_job_result` and pass an additional kwarg when executing the function.

I'm afraid I don't understand your proposed solution. Can you write some pseudo code for what the user would call when enqueueing and the job function please? If the caller accepts `rq_dependency_job_result : Dict`, doesn't that mean you need to update your job code as well? (same problem with # 1 as the decorator). 

> 1. You actually still need to update your code to use previous job results.

I agree this is a bit of an issue and I'd like to avoid it. 

What if we put the *args and **kwargs substitution code into [Queue.parse_args](https://github.com/rq/rq/blob/76ba690aafaa0fffbab1ee8f5db7c27c5327d792/rq/queue.py#L460)? We would need to unwrap it again when executing. It seems like somewhere around the [deserializer](https://github.com/rq/rq/blob/76ba690aafaa0fffbab1ee8f5db7c27c5327d792/rq/job.py#L260), although it seems strange to fetch a Job from inside another Job. I'll defer to you guys since I've only been looking at this code for a few hours.

**Comment by mattchan-tencent** [comment]
One last comment for the night: 

I think one structural weakness of the decorator method (explicit or integrated into serialize/deserialize) is that you are pinning the return type of `job1` to the parameters of `bar`. This becomes an issue when you have complex execution graphs (i.e. job1 provides parameters for both `bar` and `bar_prime`) or when you want to do some intermediate processing of parameters like renaming or reordering. 

One solution to this kind of problem might be to make Job.results return a deferred lookup key instead of None. For example:

```python 
class Results:
    def __init__(self, jid: str) 
        self.jid = jid
    
    def __getattr__(self, key):
        return f"{self.jid}:{key}"

    def __getitem__(self, index):
        return f"{self.jid}:{index}"

# ---
def foo():
    return {"x":5, "y":4}

def bar(x):
    return (2*x, 5)

# ---
j1 = q.enqueue(foo)
j2 = q.enqueue(bar, j1.results.x, depends_on=j1)
j3 = q.enqueue(bar, j2.results[0], depends_on=j2)
``` 

When the worker evaluates the lookup key, it fetches job id, then subsequently the return value of the dependency by index or key

**Comment by selwin** [comment]
> I'm afraid I don't understand your proposed solution. Can you write some pseudo code for what the user would call when enqueueing and for the job function please? If the caller accepts rq_dependency_job_result : Dict, doesn't that mean you need to update your job code as well? (same problem with # 1 as the decorator).

It doesn't solve the need to modify your job code, but does make it more explicit and easier to understand. My initial proposed solution

```python
def foo():
    return {"x":5, "y":4}

def bar(rq_job_result):
    # The argument has to be explicitly named, we can check whether function
    # accepts this argument when being executed with Python's inspect module
    # https://docs.python.org/3/library/inspect.html#module-inspect
    return (2*rq_job_result, 5)
```

But I think this is a better method (and I think we can make it work):
```python

def foo():
    return {"x":5, "y":4}

@pass_previous_job_result_to('x')
def bar(x):
    return (2*x, 5)

job1 = q.enqueue(foo)
q.enqueue(bar, job1, depends_on=job1)
```

**Comment by mattchan-tencent** [comment]
Hmm... I think this still has some problems with locality. If I wanted to change `x` from a job result to a constant, I would have to adjust the decorator on `bar` and also the `q.enqueue` call. It makes it harder to reason about my code for other programmers as well.

Also I'm assuming that `x` in the decorator is not the same x that `foo` returns in the dict. That would be a bad idea I think. 

This is the real code that I've written for testing (it works):

``` python
# job
def decorate(forig):
    @wraps(forig)
    def fnew(*args, **kwargs):
        args = [Job.fetch(a.lstrip("_reserved_"), connection=redis).result if isinstance(a, str) and a.startswith("_reserved_") else a for a in args]
        kwargs = {k:Job.fetch(v.lstrip("_reserved_"), connection=redis).result if isinstance(a, str) and a.startswith("_reserved_") else v for k,v in kwargs.items()}

        return forig(*args, **kwargs)
    return fnew

def foo():
    return 5

@decorate
def bar(x,y):
    return x * y

# ---
# submit

def wrap_queue(f):
    @wraps(f)
    def newqueue(*args, **kwargs):
        jobs = [a for a in args if isinstance(a, Job)] + [v for v in kwargs.values() if isinstance(v, Job)]
        args = [f"_reserved_{a.id}" if isinstance(a, Job) else a for a in args]
        kwargs = {k:f"_reserved_{v.id}" if isinstance(v, Job) else v for k,v in kwargs.items()}
        kwargs["depends_on"] = kwargs.get("depends_on", []) + jobs
        return f(*args, **kwargs)
    return newqueue

# ...
    new_enqueue = wrap_queue(q.enqueue)
    j1 = new_enqueue(foo)
    j2 = new_enqueue(bar, j1, 2)
```

The downsides of this approach I think is the use of runtime type inspection and the presence of a reserved key.
